src/hypothesis/internal/charmap.py

```diff
@@ -53,8 +55,13 @@ def charmap():
             data = sorted(
                 (k, tuple((map(tuple, v))))
                 for k, v in _charmap.items())
-            with gzip.GzipFile(f, 'wb', mtime=1) as o:
+
+            # Write the Unicode table atomically
+            _, tmpfile = tempfile.mkstemp()
+            with gzip.GzipFile(tmpfile, 'wb', mtime=1) as o:
                 o.write(pickle.dumps(data, pickle.HIGHEST_PROTOCOL))
```

Comment:
Maybe serialize before creating the tempfile, so we don't have to add extra code for deleting the tempfile if serialization fails.